src/useImperativeFilePicker.tsx

```diff
@@ -0,0 +1,135 @@
+import { useEffect, useState } from 'react';
+import { FileContent, ImperativeFilePickerReturnTypes, UseFilePickerConfig } from './interfaces';
+import useFilePicker from './useFilePicker';
+import { Validator } from './validators/validatorInterface';
+
+/**
+ * File limit validator has to be overriden to take into account the files that were previously selected
+ * @param previousPlainFiles files that were previously selected
+ * @returns a validator that checks if the amount of files selected previously and in the current batch is within the limits
+ */
+const persistentFileLimitValidator: (previousPlainFiles: File[]) => Validator = previousPlainFiles => ({
+  validateBeforeParsing(config: UseFilePickerConfig, plainFiles: File[]): Promise<void> {
+    const { limitFilesConfig } = config;
+    const fileAmount = previousPlainFiles.length + plainFiles.length;
+    if (limitFilesConfig) {
+      if (limitFilesConfig.max && fileAmount > limitFilesConfig.max) {
+        return Promise.reject({ maxLimitExceeded: true });
+      }
+
+      if (limitFilesConfig.min && fileAmount < limitFilesConfig.min) {
+        return Promise.reject({ minLimitNotReached: true });
+      }
+    }
+    return Promise.resolve();
+  },
+  validateAfterParsing(): Promise<void> {
+    return Promise.resolve();
+  },
+});
+
+// Comparators needed to check if the file was already selected previously
+
+const areFilesEqual = (file1: File | undefined, file2: File | undefined) =>
+  !!file1 &&
+  !!file2 &&
+  file1.name === file2.name &&
+  file1.lastModified === file2.lastModified &&
+  file1.webkitRelativePath === file2.webkitRelativePath &&
+  file1.size === file2.size &&
+  file1.type === file2.type;
+
+const areFilesContentEqual = (file1: FileContent | undefined, file2: FileContent | undefined) =>
+  !!file1 &&
+  !!file2 &&
+  file1.name === file2.name &&
+  file1.lastModified === file2.lastModified &&
+  file1.size === file2.size &&
+  file1.type === file2.type &&
+  file1.content === file2.content;
+
+/**
+ * A version of useFilePicker hook that keeps selected files between selections. On top of that it allows to remove files from the selection.
+ */
+function useImperativeFilePicker(props: UseFilePickerConfig): ImperativeFilePickerReturnTypes {
+  const { readFilesContent, onFilesSelected, onFilesSuccessfulySelected } = props;
+
+  const [allPlainFiles, setAllPlainFiles] = useState<File[]>([]);
+  const [allFilesContent, setAllFilesContent] = useState<FileContent[]>([]);
+
+  const [open, { plainFiles, filesContent, loading, errors, clear }] = useFilePicker({
+    ...props,
+    validators: [persistentFileLimitValidator(allPlainFiles), ...(props.validators || [])],
```

Comment:
Can we only validate the new one? If the old one are existing then it's already enough of a check right?

Reply:
For every other validator it should work, but for the file limit validator that isn't the case. Without that fix, when you limit the amount of files to be selected at once to 2, then you'd be able to select 2 files multiple times which will be appended to the all files array, so the user can easily bypass the limit amount check.